examples/custom_rule.jl

```diff
@@ -0,0 +1,191 @@
+# To squeeze the last drop of performance, the below rule avoids computing the output of the original function and 
+# just computes its derivative.
+
+function forward(func::Const{typeof(f)}, ::Type{<:DuplicatedNoNeed}, y::Duplicated, x::Duplicated)
+    println("Using custom rule with DuplicatedNoNeed output.")
+    y.val .= x.val.^2 
+    y.dval .= 2 .* x.val .* x.dval
+    return sum(y.dval)
+end
+
+# Our rule is triggered, for example, when we call `autodiff` directly on `f`, as the output's derivative isn't needed:
+
+x  = [3.0, 1.0]
+dx = [1.0, 0.0]
+y  = [0.0, 0.0]
+dy = [0.0, 0.0]
+
+@show autodiff(Forward, f, Duplicated(y, dy), Duplicated(x, dx)) # derivative of f w.r.t. x[1]
+@show dy; # derivative of y w.r.t. x[1] when f is run
+
+# !!! note "Custom rule dispatch"
+#     When multiple custom rules for a function are defined, the correct rule is chosen using 
+#     [Julia's multiple dispatch](https://docs.julialang.org/en/v1/manual/methods/#Methods).
+#     In particular, it is important to understand that the custom rule does not *determine* the
+#     activities of the inputs and the outputs: rather, `Enzyme` decides the activity annotations independently,
+#     and then *dispatches* to the custom rule handling the activities, if one exists.
+#     If a custom rule is specified for the correct function/argument types, but not the correct activity annotation, 
+#     a runtime error will be thrown alerting the user to the missing activity rule rather than silently ignoring the rule."
+
+# Finally, it may be that either `x` or `y` are marked as [`Const`](@ref). We can in fact handle this case, 
+# along with the previous two cases, all together in a single rule:
+
+function forward(func::Const{typeof(f)}, RT::Type{<:Union{DuplicatedNoNeed, Duplicated}}, 
+                 y::Union{Const, Duplicated}, x::Union{Const, Duplicated})
+    println("Using our general custom rule!")
+    y.val .= x.val.^2 
+    if !(x <: Const) && !(y <: Const)
+        y.dval .= 2 .* x.val .* x.dval
+    elseif !(y <: Const) 
+        y.dval .= 0
+    end
+    if RT <: DuplicatedNoNeed
+        return sum(y.dval)
+    else
+        return Duplicated(sum(y.val), sum(y.dval))
+    end
+end
+
+# Note that there are also exist batched duplicated annotations for forward mode, namely [`BatchDuplicated`](@ref)
+# and [`BatchDuplicatedNoNeed`](@ref), which are not covered in this tutorial.
+
+# ## Defining a reverse-mode rule
+
+# Finally, let's look at how to write a simple reverse-mode rule! 
+# First, we write a method for [`EnzymeRules.augmented_primal`](@ref):
+
+function augmented_primal(config::ConfigWidth{1}, func::Const{typeof(f)}, ::Type{<:Active},
+                          y::Duplicated, x::Duplicated)
+    println("In custom augmented primal rule.")
+    if needs_primal(config)
+        return AugmentedReturn(func.val(y.val, x.val), nothing, nothing)
+    else
+        return AugmentedReturn(nothing, nothing, nothing)
```

Comment:
This isn't right since it won't set y in this case whereas the original code would.